src/GovUk.Education.ExploreEducationStatistics.Data.Services/SubjectMetaService.cs

```diff
@@ -32,91 +34,128 @@ private enum SubjectMetaQueryStep
             GetFilterItems
         }
 
-        private readonly StatisticsDbContext _context;
+        private readonly IPersistenceHelper<StatisticsDbContext> _persistenceHelper;
+        private readonly StatisticsDbContext _statisticsDbContext;
         private readonly IFilterRepository _filterRepository;
+        private readonly IFilterItemRepository _filterItemRepository;
         private readonly IIndicatorGroupRepository _indicatorGroupRepository;
         private readonly ILocationRepository _locationRepository;
         private readonly ILogger _logger;
         private readonly IObservationService _observationService;
-        private readonly IPersistenceHelper<StatisticsDbContext> _persistenceHelper;
         private readonly ITimePeriodService _timePeriodService;
         private readonly IUserService _userService;
         private readonly LocationsOptions _locationOptions;
 
-        public SubjectMetaService(
-            StatisticsDbContext context,
+        public SubjectMetaService(IPersistenceHelper<StatisticsDbContext> persistenceHelper,
+            StatisticsDbContext statisticsDbContext,
             IFilterRepository filterRepository,
             IFilterItemRepository filterItemRepository,
             IIndicatorGroupRepository indicatorGroupRepository,
             ILocationRepository locationRepository,
             ILogger<SubjectMetaService> logger,
             IObservationService observationService,
-            IPersistenceHelper<StatisticsDbContext> persistenceHelper,
             ITimePeriodService timePeriodService,
             IUserService userService,
-            IOptions<LocationsOptions> locationOptions) :
-            base(filterItemRepository)
+            IOptions<LocationsOptions> locationOptions)
         {
-            _context = context;
+            _persistenceHelper = persistenceHelper;
+            _statisticsDbContext = statisticsDbContext;
             _filterRepository = filterRepository;
+            _filterItemRepository = filterItemRepository;
             _indicatorGroupRepository = indicatorGroupRepository;
             _locationRepository = locationRepository;
             _logger = logger;
             _observationService = observationService;
-            _persistenceHelper = persistenceHelper;
             _timePeriodService = timePeriodService;
             _userService = userService;
             _locationOptions = locationOptions.Value;
         }
 
-        public Task<Either<ActionResult, SubjectMetaViewModel>> GetSubjectMeta(Guid subjectId)
+        public async Task<Either<ActionResult, SubjectMetaViewModel>> GetSubjectMeta(ReleaseSubject releaseSubject)
```

Comment:
Not sure if I'm missing something here but how come we've gone with passing in `ReleaseSubject` as a parameter? Kinda would have expected still expected a `subjectId` and optional `releaseId` so that we could fetch the `ReleaseSubject` within the service.

Ideally it'd be good to avoid passing around entities as much as possible. There's obviously caveats around whether relationships have been loaded beforehand or not. It'd be nice to keep as much of the release subject fetching in this service too.

If this mostly due to the cache key stuff, perhaps it'd be good to have overloads for `ReleaseSubject` and `releaseId` / `subjectId`.

Same applies for `FilterSubjectMeta`.

Reply:
Yes, some requests don't have a `releaseId` so we need to lookup the `ReleaseSubject` for the latest version of the release which could all be done in the service if it wasn't for caching.

Prior to your comment I made the `GetSubjectMeta` and `FilterSubjectMeta` methods in the service take `ReleaseSubject` for consistency and updated everything across both the Admin and Data.API, thinking this would be more readable and have less variations of tests plus permissions tests.

I've applied your suggestion now and altered the Admin controller to always use `releaseId` and `subjectId` for both `GetSubjectMeta` and `FilterSubjectMeta`. In the Admin `releaseId` is always known. This tidies things up nicely for that controller.

In the Data.API it now always calls the service with `releaseId?` and `subjectId` for `FilterSubjectMeta` but still uses a `releaseSubject` overload for `GetFilterSubjectMeta` due to the caching.

For the caching, in order to know the release slug we can use `ReleaseSubject.Release.Slug` but because we also need to know the publication slug we lookup the Content release including publication and then we have both. Either way we need to have identified the release from `subjectId`.

We could potentially do away with wrapping the `ReleaseSubject` in the cacheable entity as long as it still held the required parameters for caching, i.e. both slugs and the subject id. However as the ReleaseSubject has already been fetched I've decided to continue using the overload.

Just as an aside, for the case where we don't have the `releaseId` I've been contemplating changing it so that rather than looking up the latest ReleaseSubject using `subjectId` with `_releaseSubjectRepository.GetReleaseSubjectForLatestPublishedVersion`, I could lookup the publication from any ReleaseSubject using the `subjectId` along with looking up the latest release in the series for that publication. Then call the service with `subjectId` and the latest release in the series' `releaseId`.

Maybe this is the more 'correct' thing to do as it would stop someone fetching the metadata without a release id for a subject of a release that's no longer the latest release in the series.

At the moment you would be able to query without release id for a subject  of a release that's no longer the latest release in the series. 

It must always be the latest version of that release however, and that's enforced with a permission check on the service, so I've not made this change for now.

Reply:
Fair enough, sounds reasonable.

As to the final point about how we fetching the latest published release, I'm not sure I understand the following:

> Maybe this is the more 'correct' thing to do as it would stop someone fetching the metadata without a release id for a subject of a release that's no longer the latest release in the series.

How would someone be able to get the meta for the non-latest release without a `releaseId`? I would have assumed `_releaseSubjectRepository.GetReleaseSubjectForLatestPublishedVersion` should already prevent that?

Reply:
Short answer: It's allowing them to get it for the latest version of all published releases, which I think is ok? 🙂 

Long answer:
`_releaseSubjectRepository.GetReleaseSubjectForLatestPublishedVersion ` works by assuming a subject can be linked to more than one release through `ReleaseSubject` but working out which is the most recent release version by dismissing any older versions that are linked through `previousVersionId` and dismissing any where the release is not `live`. This ends up with either one result or no result.

It's not aware of content releases and that there may be a newer published release _in the series_ for a later time period.  I put 'version' in the name of the method to try and make it a bit clearer what it's returning.

Getting the meta data either with a specific `releaseId` from a previous _version_ that's been replaced, or by `subjectId` alone where the subject has been removed from the latest _version_ will both be forbidden by the permissions check on `DataSecurityPolicies.CanViewSubjectData` which checks the content release is the latest published _version_. I think this is the important bit, i.e. that we don't allow returning meta data where the specified release id, or the releasesubject that's been found aren't on a published release.

Getting the meta data either with a specific `releaseId` from the latest version or by `subjectId` alone where the subject is on the latest version, but in both cases where that version is not the latest release in the series by time period will both be allowed because a ReleaseSubject will be found and the `DataSecurityPolicies.CanViewSubjectData` is happy because that release is published, just not the latest year/time period.